libdevice/imf/exp_h_la.cpp

```diff
@@ -0,0 +1,125 @@
+} __devicelib_imf_internal_hexp_data_t;
+static const __devicelib_imf_internal_hexp_data_t
+    __devicelib_imf_internal_hexp_data = {
+        0x0100u, //__Sbit
+        0x5300u, //__Shifter
+        0x3dc5u, //__Log2E
+        0xb98Cu, //__LN2
+        0x0af4u, //__LN2L
+        {
+            //__Tbl_exp
+            0x3c00u, 0x3c16u, 0x3c2du, 0x3c45u, 0x3c5du, 0x3c75u, 0x3c8eu,
+            0x3ca8u, 0x3cc2u, 0x3cdcu, 0x3cf8u, 0x3d14u, 0x3d30u, 0x3d4du,
+            0x3d6bu, 0x3d89u, 0x3da8u, 0x3dc8u, 0x3de8u, 0x3e09u, 0x3e2bu,
+            0x3e4eu, 0x3e71u, 0x3e95u, 0x3ebau, 0x3ee0u, 0x3f06u, 0x3f2eu,
+            0x3f56u, 0x3f7fu, 0x3fa9u, 0x3fd4u,
+        },
+        0xbfffu, //__Rmask
+        0x7fffu, //__Absmask
+        0xcc80u, //__M18
+        0x4a00u, //__P12
+        0x6a0fu, //__Shifter2
+        0x3dc5u, //__L2E
+        0x0d1eu, //__L2EL
+        0x2110u, //__gc3
+        0x2b52u, //__gc2
+        0x33afu, //__gc1
+        0x398bu, //__gc0
+        0x7c00u, //__Inf
+};
+
+static inline _iml_half __devicelib_imf_expf16_impl(_iml_half a) {
+  VUINT32 vm;
+  _iml_half va1;
+  _iml_half vr1;
+  _iml_half r;
+  va1 = a;
+  ;
+  {
+    _iml_half X;
+    _iml_half P12;
+    _iml_half M18;
+    _iml_half L2E;
+    _iml_half L2EL;
+    _iml_half Shifter;
+    _iml_half Y;
+    _iml_half N;
+    _iml_half PN2;
+    _iml_half Poly;
+    _iml_half Inf;
+    VUINT16 NanMask;
+    _iml_half R;
+    _iml_half c3;
+    _iml_half c2;
+    _iml_half c1;
+    _iml_half c0;
+    vm = 0;
+    // restrict input range to [-18.0,12.0]
+    P12 = as_half(__devicelib_imf_internal_hexp_data.__P12);
+    M18 = as_half(__devicelib_imf_internal_hexp_data.__M18);
+    X = ((va1 < P12) ? va1 : P12);
+    X = ((X > M18) ? X : M18);
+    // (2^11*1.5 + bias) + x*log2(e)
+    Shifter = as_half(__devicelib_imf_internal_hexp_data.__Shifter2);
+    L2E = as_half(__devicelib_imf_internal_hexp_data.__L2E);
+    Y = __fma(X, L2E, Shifter);
+    // N = 2*(int)(x*log2(e)/2)
+    N = (Y - Shifter);
+    // 2^(N/2)
+    PN2 = as_half(((VUINT16)as_short(Y) << (10)));
+    // reduced arg.: x*log2(e) - N
+    R = __fma(X, L2E, -(N));
+    L2EL = as_half(__devicelib_imf_internal_hexp_data.__L2EL);
+    R = __fma(X, L2EL, R);
+    // polynomial ~ 2^R
+    c3 = as_half(__devicelib_imf_internal_hexp_data.__gc3);
+    c2 = as_half(__devicelib_imf_internal_hexp_data.__gc2);
+    c1 = as_half(__devicelib_imf_internal_hexp_data.__gc1);
+    c0 = as_half(__devicelib_imf_internal_hexp_data.__gc0);
+    // start polynomial
+    Poly = __fma(c3, R, c2);
+    Poly = __fma(Poly, R, c1);
+    Poly = __fma(Poly, R, c0);
+    Poly = (Poly * R);
+    vr1 = __fma(Poly, PN2, PN2);
+    // fixup for input=NaN
+    Inf = as_half(__devicelib_imf_internal_hexp_data.__Inf);
+    NanMask = (!(va1 <= Inf)) ? 0x1L : 0x0L;
+    PN2 = (NanMask != 0x0) ? (va1) : (PN2);
+    // result:  (1+poly)*2^(N/2)*2^(N/2)
+    vr1 = (vr1 * PN2);
+  }
+  r = vr1;
+  return r;
+}
+
+DEVICE_EXTERN_C_INLINE
```

Comment:
While I believe that suggested approach where each fp16 arithmetic operation is implemented through fp16->fp32->op32->fp16 conversion should work but most possibly it will be presumably slower than if we just convert 16->32, call 32-bit math function as a whole and convert 32 result back to 16: fp16->fp32->exp32->fp16.
So, if we do care about host performance in fp16 case than the second approach looks more preferable on hosts with no fp16 hardware support. Host with fp16 hardware support should map all fp16 arithmetic operations to native 16-bit floating point instructions.

Reply:
Hi, @akolesov-intel 
I understand fp16->fp32->exp32->fp16 should have better perf than fp16->fp32->op32->fp16. The intention to do "fp16->fp32->op32->fp16" is to share the same implementation  between host device and other device. 
For SYCL host device, performance is not so important, we only need to guarantee functionality since the most important usage for SYCL host device is to provide a reference for other devices(CPU, GPU, FPGA).
Thanks very much.

Reply:
That's fine if performance is not important for host